### ISSUE TITLE
Wait for `probe` not for `result`

### ISSUE THREAD

**Issue by shaardie** [lkiesow/go-scan-http#7]
Currently you are counting all results and stop if they are present. I would suggest better wait on all probes to finish and gather all results. If a probe for example stops without result, you will never finish. You could use closures and a waitgroup for that. Like

```go
var wg sync.WaitGroup
for b0 := settings.bytes[0][0]; b0 <= settings.bytes[0][1]; b0++ {
	for b1 := settings.bytes[1][0]; b1 <= settings.bytes[1][1]; b1++ {
		for b2 := settings.bytes[2][0]; b2 <= settings.bytes[2][1]; b2++ {
			for b3 := settings.bytes[3][0]; b3 <= settings.bytes[3][1]; b3++ {
				for _, port := range settings.ports {
					addr := fmt.Sprintf("%d.%d.%d.%d:%d",
						b0, b1, b2, b3, port)
					maxqueue <- true
					go func() {
						wg.Add(1)
						probe(addr, results, maxqueue)
						wg.Done()
						}()
				}
			}
		}
	}
}
wg.Wait()
```
... not tested.

**Comment by lkiesow** [comment]
That would mean buffering up to 2^32 messages. That's a lot of memory ;-)

`probe(…)` already ensures that it's writing to the `results` channel but it might be cleaner if `probe(…)` would return a result and an error and the error handling and channel feeding happens right in the main loop in a go routine.

Do you know how much overhead an additional inline function has in go? E.g. `go func() {probe(…)}` vs `go probe(…)`?

**Comment by shaardie** [comment]
We do you get the 2^32 messages buffering from?

How do you "ensure" that `probe(...)` is writing to the channel? If it is only ensured by the program logic itself, you can easily create a deadlock of the program in the future. Why do you want to count the result when printing them anyway? What is the gain? But I like the idea of returning a result + error and handle it elsewhere :+1:

My hopes are on the compiler for performance, but I actually do not know.

**Comment by lkiesow** [comment]
> We do you get the 2^32 messages buffering from?
>
> > I would suggest better wait on all probes to finish and gather all results.

The IPv4 rage is a 32-bit integer. Hence iterating over the full range will give you 2³² × number of scanned ports results. Right now, it's ensured that only a specific number of results are in memory. Waiting on all results… well, you then have to put them all in memory.

> How do you "ensure" that `probe(...)` is writing to the channel? If it is only ensured by the program logic itself, you can easily create a deadlock of the program in the future.

That is a universal truth ;-D
Right now, your suggested change would even result in one since `results` is a buffered channel and you would not read from it anymore until all results are collected. Something which might never happen if there are too many results.

In other words: As soon as you start doing things concurrently, your program's logic has to ensure no deadlocks occur.

> Why do you want to count the result when printing them anyway? What is the gain?

Your change does the same thing although you use a different syntax. The WaitGroup ensures that we wait for all go routines *by counting them*. Instead of continuously calling `eg.Add(1)` you could also just call `wg.Add(<number of requests>)` once before the loop. There is no difference, except that you end up with fewer function calls.

Still, using a WaitGroup might be nicer than doing it on your own. I've changed that and pushed the error handling out of `probe(…)`.
